mmap-alloc/src/lib.rs

```diff
@@ -295,6 +295,56 @@ impl MapAlloc {
         map(size, self.perms, self.commit, self.huge_pagesize).and_then(f)
     }
 
+    #[cfg(target_os = "linux")]
+    unsafe fn realloc_helper(&self, old_ptr: *mut u8, old_size: usize, new_size: usize) -> Option<*mut u8> {
+        // Since allocators in Rust are not allowed to return null pointers, but it may be valid for
+        // mremap to return memory starting at null, we have to handle that case. We
+        // do this by checking for null, and if we find that map has returned null, we create a
+        // new mapping, and mremap the data onto that new mapping, and try to map a new empty page
+        // at address 0. Since we leave the page at 0 mapped, future calls should never return null.
+        // Note that this leaks memory since we never unmap that page, but this isn't a big deal -
+        // since we never write to it, it will remain uncommitted and will thus not consume any
+        // physical memory.
+        let f = |ptr: *mut u8| if ptr.is_null() {
+            let new_ptr = map(new_size, self.perms, false, self.huge_pagesize)
```

Comment:
Add a comment like "first create a mapping that will serve as the destination of the remap".